src/sec.c

```diff
@@ -0,0 +1,263 @@
+    size_t old_size, size_t new_size, bool zero);
+static bool sec_shrink(tsdn_t *tsdn, pai_t *self, edata_t *edata,
+    size_t old_size, size_t new_size);
+static void sec_dalloc(tsdn_t *tsdn, pai_t *self, edata_t *edata);
+
+bool sec_init(sec_t *sec, pai_t *fallback, size_t nshards, size_t alloc_max,
+    size_t bytes_max) {
+	if (nshards > SEC_NSHARDS_MAX) {
+		nshards = SEC_NSHARDS_MAX;
+	}
+	for (size_t i = 0; i < nshards; i++) {
+		sec_shard_t *shard = &sec->shards[i];
+		bool err = malloc_mutex_init(&shard->mtx, "sec_shard",
+		    WITNESS_RANK_SEC_SHARD, malloc_mutex_rank_exclusive);
+		if (err) {
+			return true;
+		}
+		shard->enabled = true;
+		for (pszind_t j = 0; j < SEC_NPSIZES; j++) {
+			edata_list_active_init(&shard->freelist[j]);
+		}
+		shard->bytes_cur = 0;
+	}
+	sec->fallback = fallback;
+	sec->alloc_max = alloc_max;
+	if (sec->alloc_max > sz_pind2sz(SEC_NPSIZES - 1)) {
+		sec->alloc_max = sz_pind2sz(SEC_NPSIZES - 1);
+	}
+
+	sec->bytes_max = bytes_max;
+	sec->nshards = nshards;
+
+	/*
+	 * Initialize these last so that an improper use of an SEC whose
+	 * initialization failed will segfault in an easy-to-spot way.
+	 */
+	sec->pai.alloc = &sec_alloc;
+	sec->pai.expand = &sec_expand;
+	sec->pai.shrink = &sec_shrink;
+	sec->pai.dalloc = &sec_dalloc;
+
+	return false;
+}
+
+static sec_shard_t *
+sec_shard_pick(tsdn_t *tsdn, sec_t *sec) {
+	/*
+	 * Eventually, we should implement affinity, tracking source shard using
+	 * the edata_t's newly freed up fields.  For now, just randomly
+	 * distribute across all shards.
+	 */
+	if (tsdn_null(tsdn)) {
+		return &sec->shards[0];
+	}
+	tsd_t *tsd = tsdn_tsd(tsdn);
+	/*
+	 * Use the trick from Daniel Lemire's "A fast alternative to the modulo
+	 * reduction.  Use a 64 bit number to store 32 bits, since we'll
+	 * deliberately overflow when we multiply by the number of shards.
+	 */
+	uint64_t rand32 = prng_lg_range_u64(tsd_prng_statep_get(tsd), 32);
+	uint32_t idx = (uint32_t)((rand32 * (uint64_t)sec->nshards) >> 32);
+	return &sec->shards[idx];
+}
+
+static edata_t *
+sec_shard_alloc_locked(tsdn_t *tsdn, sec_t *sec, sec_shard_t *shard,
+    pszind_t pszind) {
+	malloc_mutex_assert_owner(tsdn, &shard->mtx);
+	if (!shard->enabled) {
+		return NULL;
+	}
+	edata_t *edata = edata_list_active_first(&shard->freelist[pszind]);
+	if (edata != NULL) {
+		edata_list_active_remove(&shard->freelist[pszind], edata);
+		assert(edata_size_get(edata) <= shard->bytes_cur);
+		shard->bytes_cur -= edata_size_get(edata);
+	}
+	return edata;
+}
+
+static edata_t *
+sec_alloc(tsdn_t *tsdn, pai_t *self, size_t size, size_t alignment, bool zero) {
+	assert((size & PAGE_MASK) == 0);
+
+	sec_t *sec = (sec_t *)self;
+
+	if (zero || alignment > PAGE || sec->nshards == 0
+	    || size > sec->alloc_max) {
+		return pai_alloc(tsdn, sec->fallback, size, alignment, zero);
+	}
+	pszind_t pszind = sz_psz2ind(size);
+	sec_shard_t *shard = sec_shard_pick(tsdn, sec);
+	malloc_mutex_lock(tsdn, &shard->mtx);
```

Comment:
maybe do an atomic load on the bytes, and skip the whole lock-alloc-unlock if it's 0?

Reply:
This is probably not compatible with later batching optimizations -- we'll want to hold the lock for a more complicated operation than just examining a linked list. (It also shouldn't fire often -- if it does, it means that caching in general is failing, in which case micro-optimizing the sharded layer is the least of our worries).